### PR TITLE
[RDY] Customize Consume Item menu

### PR DESCRIPTION
**Pull request by codemime**
Follows #19952 and #18681.

### Overview

- **The usable information and various reasons why you can't consume stuff are shown.**

- **CBMs that turn items into energy became more UI friendly:**

- **All the "queries" are now asked at once. No more annoying repetition of "Really eat that?":**
reasons

- **"Smart" sorting order:**

   - Rotten food goes last unless you're a saprophagous (then it goes first).
   - Perishable food goes before long-lasting one (preservation and freshness were taken into account).   
   - ~~If you're thirsty, food gets sorted by QUENCH, otherwise by NUTRITION.~~
   - ~~Then by what remains (either NUTRITION or QUENCH).~~
   - ~~Then by JOY.~~
   - ~~Then by ENERGY (in case of active CBMs).~~

### CODE REVIEW

File: src/consumption.cpp [hunk 1]
```diff
@@ -215,191 +227,157 @@ morale_type player::allergy_type( const item &food ) const
     return MORALE_NULL;
 }
 
-bool player::is_allergic( const item &food ) const
+edible_rating player::can_eat( const item &food, std::string *err ) const
```
Comment (Coolthulhu) [inline src/consumption.cpp]: The interface would be cleaner if the return value included the error - return by parameter is C-ish.
For example, if you returned it as a `const std::pair<edible_rating, std::string> &`.

**Comment by Coolthulhu** [comment]
I'm not sure if that sorting is a good idea, at least not without a way to turn it off.
One very important quality of menus is that they do not change if they don't need to: that they keep dimensions and contents as consistent as possible.

Is there a different item menu that is sorted by default? If not, this one should only be sorted once the player presses some sort/filter button.

File: src/consumption.cpp (hunk 1 above)
Comment (codemime) [inline src/consumption.cpp]: > The interface would be cleaner if the return value included the error - return by parameter is C-ish.

Agreed. For now, I've simply repeated the signature of player::can_eat

> For example, if you returned it as a const std::pair<edible_rating, std::string> &

I was thinking about that too and came up with a template class. Something like this:

```c++
template<typename T = bool, T no_error = true, T default_error = false>
class ret_val
{
    public:
        static_assert( no_error != default_error, "Error codes must differ." );

        static ret_val success( const std::string &msg = std::string() ) {
            return ret_val( no_error, msg );
        }

        static ret_val failure( T code = default_error, const std::string &msg = std::string() ) {
            return ret_val( default_error, msg );
        }

        operator bool() const {
            return code == no_error;
        }

        T operator *() const {
            return code;
        }

        const std::string &get_msg() const {
            return msg;
        }

    private:
        ret_val( T code, const std::string &msg ) : code( code ), msg( msg ) {}

    private:
        std::string msg;
        T code;
};

```
I think this should be done in a separate PR for all related cases.

**Comment by codemime** [comment]
> I'm not sure if that sorting is a good idea, at least not without a way to turn it off.

Neither am I. That's mostly why I put the CR tag.

> One very important quality of menus is that they do not change if they don't need to: that they keep dimensions and contents as consistent as possible.

They'd be more stable if sorted by NUTRITION only. Or even solely by freshness (as an extreme). I do think that the latter is a good idea. 

> Is there a different item menu that is sorted by default? 

Gunmod menu is sorted using the installation odds. 

> If not, this one should only be sorted once the player presses some sort/filter button.

As for now, there's no way we can toggle sorting. This probably needs to be implemented.

**Comment by Coolthulhu** [comment]
The display is generally fine on a sane setting, but our defaults are not sane: at 80 screen width, the new display columns force inventory columns (from map, from vehicle etc.) below the inventory ones, halving the number of items displayed.
Now, 80 screen width is generally rather painful and I wouldn't recommend playing with such a narrow viewport, but it is the default and thus we're supposed to support it.
Is there a way to compress the display to force the 2 columns if there is only one?

If not, I'll wait for others to approve. It's fine by me, but not necessarily for everyone else.

As for the smart sorting:
This one sounds like it could be painful in some cases (finding specific joy item in a huge stash etc.). I'd add an option (as in, options.cpp option) to disable it, at least at first. It can be the default, just that it shouldn't be the only option until we can *sort* it all out.

One thing could really help with everything here:
Being able to filter the inventory in a similar way all `uimenu`s are filtered. If this was implemented, the sorting would be totally fine in pretty much all cases.

**Comment by codemime** [comment]
> Is there a way to compress the display to force the 2 columns if there is only one?

Doable, but I doubt that makes sense. Long rows would either overlap each other or simply won't fit. BTW I didn't change column management in this PR.

> This one sounds like it could be painful in some cases.  I'd add an option (as in, options.cpp option) to disable it, at least at first. It can be the default, just that it shouldn't be the only option until we can sort it all out.

That would be equally painful with lists sorted alphabetically, but the option is an _option_.

> Being able to filter the inventory in a similar way all uimenus are filtered.

I thought about that too and I'm going to implement that.

**Comment by codemime** [comment]
I have resolved the merge conflict and removed all additional sorting except by freshness.

File: src/consumption.cpp [hunk 2]
```diff
@@ -205,191 +217,159 @@ morale_type player::allergy_type( const item &food ) const
...
     const bool saprophage = has_trait( trait_id( "SAPROPHAGE" ) );
-    if( spoiled ) {
-        if( !saprophage && !has_trait( trait_id( "SAPROVORE" ) ) &&
-            !maybe_query( _( "This %s smells awful!  Eat it?" ) ) ) {
-            return ROTTEN;
+
+    if( food.rotten() ) {
+        const bool saprovore = has_trait( trait_id( "SAPROVORE" ) );
+        if( !saprophage && !saprovore ) {
+            consequences.emplace_back( ROTTEN, _( "This is rotten an smells awful!" ) );
```
Comment (Coolthulhu) [inline src/consumption.cpp]: Typo

File: src/consumption.cpp [hunk 3]
```diff
@@ -205,191 +217,159 @@ morale_type player::allergy_type( const item &food ) const
...
-            overfull = !maybe_query( _( "You will not be able to finish it all.  Consume it?" ) );
+    const int nutr = nutrition_for( food.type );
+    const int quench = comest->quench;
+    const int temp_hunger = get_hunger() - nutr;
+    const int temp_thirst = get_thirst() - quench;
+
+    if( has_active_mutation( trait_id( "HIBERNATE" ) ) ) {
+        if( ( get_hunger() >= -60 && get_thirst() >= -60 ) &&
+            ( temp_hunger < -60 || temp_thirst < -60 ) ) {
+            consequences.emplace_back( TOO_FULL, _( "You're adequately fueled. Prepare for hibernation?" ) );
```
Comment (Coolthulhu) [inline src/consumption.cpp]: What will be the context of this message?
You removed questions from other messages, why not this one?

File: src/consumption.cpp [hunk 4]
```diff
@@ -205,191 +217,159 @@ morale_type player::allergy_type( const item &food ) const
...
+            return res;
+        }
+        std::ostringstream req;
+        for( const auto &elem : consequences ) {
+            req << elem.second << std::endl;
+        }
+        req << _( "Consume anyway?" );
+
+        if( !query_yn( "%s", req.str().c_str() ) ) {
+            return res;
```
Comment (Coolthulhu) [inline src/consumption.cpp]: Why return the first one? Does it matter?

File: src/player.h [hunk 5]
```diff
@@ -765,20 +772,34 @@ class player : public Character, public JsonSerializer, public JsonDeserializer
...
-        /** This block is to be moved to character.h */
-        bool is_allergic( const item &food ) const;
         /** Returns allergy type or MORALE_NULL if not allergic for this player */
         morale_type allergy_type( const item &food ) const;
         /** Used for eating entered comestible, returns true if comestible is successfully eaten */
         bool eat( item &food, bool force = false );
-        edible_rating can_eat( const item &food, bool interactive = false,
-                               bool force = false ) const;
+
+        /** The food can be (theoretically) eaten. No matter the consquences. */
```
Comment (Coolthulhu) [inline src/player.h]: Is this regarding the return value?
Would be cleaner with "Can" in front.

File: src/consumption.cpp (hunk 3 above)
Comment (codemime) [inline src/consumption.cpp]: I probably wasn't careful enough during merging.

File: src/consumption.cpp (hunk 4 above)
Comment (codemime) [inline src/consumption.cpp]: It matter mostly for the colors in the inventory menu. Another option is choosing the most severe out of multiple choices.

Comment (codemime) [inline src/consumption.cpp]: I have reordered the consequences by severity (the order is arguable):

Type | Consequences
--- | ---
Rotten		| Food poisoning, health damage
Cannibalism	| Long lasting morale penalty
Nausea		| Loosing eaten food, hunger, thirst, morale penalty
Allergy		| Morale penalty that lasts half as long as for cannibalism
Too full		| Risk of sudden vomiting
Won't eat	| No ill effects

**Comment by Coolthulhu** [comment]
That prompt for "really burn x for energy" is really annoying. I assume it is to prevent burning weapons and such, but this needs to be special cased for "quality" items, otherwise it will lead to common annoyance to prevent a rare edge case.
I don't see a good way of sneaking that prompt in somehow.

Also double prompt when eating something burnable for fuel that is edible. Double Y/N queries (other than "are you really, really sure? this will probably have horrible results") are pretty much always wrong and can only ever be lesser evil, not the proper solution.
After removing the "really burn" prompt, this could be fixed by simply not allowing burning edible things. Edible things are generally bottles and boxes - if someone really wants to eat one, they'll empty it and then eat the bottle, not burn a bottle filled with water.

When the furnace is turned on, burnable food containers are colored as if inedible.
Would be fixed by simply not treating them as burnable.

**Comment by Coolthulhu** [comment]
Also a bug, but I'm not sure if it's introduced here:
* Spawn contained fast food french fries
* Drop all worn containers so that your free carrying capacity is zero
* Try to eat the fries

This will result in "Too big to pick up", even though you are trying to eat them, not hold them in the inventory.
Eating isn't a long action that requires eaten items to reside in inventory. Eating from the table is perfectly fine.

**Comment by Coolthulhu** [comment]
Fails bionic test:
```
bionics_test.cpp:56: FAILED:

  REQUIRE( p.can_consume( it ) == when_none )

with expansion:

  false == true

with messages:

  no power capacity at first

  adding Power Storage CBM only increases capacity

  'battery' is count-by-charges

  consume 'battery' with 0 charges
```

**Comment by codemime** [comment]
The test was failing because it's assumed that discharged batteries can still be used to recharge the power storage. Seems quite impossible to reproduce for the `count_by_charges` item, so no big deal.

**Comment by codemime** [comment]
All set.
